### PR TITLE
ML-507 Gaussian Process Regression

### PR DESCRIPTION
**Pull request by lilyclemson**
Random Fourier Features accelerated Gaussian Process Regressor

### CODE REVIEW

File: GPRI.ecl [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      RETURN sessID;
+  END;
+  /**
+    * Train a RFF acclerated GPR model
+    * @params session session ID for the training process.
+    * @params x independent training data.
+    * @params y dependent training data.
+    * @params rff_dim dimesion of random fourier features.
+    * @params sigma squre root of the variance.
+    * @return rst GPR model
```
Comment (RogerDev) [inline GPRI.ecl:56]: Where is GPR model defined.  Should be noted here.

File: GPRI.ecl [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      dy := SORT(DISTRIBUTE(y, 0), id, number, LOCAL);
+      rst := rffGPR.fit(session, dx, dy, rff_dim, sigma);
+      RETURN rst;
+  END;
+  /**
+    * Predict using trained GPR model
+    * @params session session ID for the predicting process.
+    * @params mod trained GPR model.
+    * @params x input data for prediction.
+    * @return rst prediction result
```
Comment (RogerDev) [inline GPRI.ecl]: Leave name out of @return statements.  Include expected format..  Should be (e.g., "@return prediction result in NumericField format")

File: README.md [hunk 3]
```diff
@@ -1 +1,2 @@
-# GaussianProcessRegression
\ No newline at end of file
+# GaussianProcessRegression
+Random Fourier Features accelerated Gaussian Process Regressor
```
Comment (RogerDev) [inline README.md]: Provide an overview of the bundle.  Specifically, The purpose and constraints of the bundleshould be discussed, and the available "public" modules should be defined and described.  For example:

GPRI is the main module that allows fitting of models and making predictions based on the fitted model.
Types provides ...
Usage examples can be found in the Test folder
etc.

Need to acknowledge the python module that we built it from.

Installation instructions:

Do we have any installation dependencies?
ecl bundle install ....

File: eclcc.log [hunk 4]
```diff
@@ -0,0 +1,38 @@
+00000000 PRG 2022-01-12 12:39:17 26888  4004 "Loading plugin C:\Program Files\HPCCSystems\8.2.18\clienttools\plugins\auditlib.dll[lib_auditlib] version = AUDITLIB 1.0.1"
```
Comment (RogerDev) [inline eclcc.log]: This file should be removed, and supress via .gitignore

File: GPRI.ecl [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+IMPORT $.Types as Types;
+IMPORT $.internal.rffGPR as rffGPR;
+
+// Common record structures
+layout_model2 := MTypes.layout_model2;
+nNodes := Thorlib.nodes();
+nodeId := Thorlib.node();
+initParams := Types.initParams;
+NumericField := MTypes.NumericField;
+
```
Comment (RogerDev) [inline GPRI.ecl:13]: Doesn't look like this bundle supports the Myriad interface.  We should note whether or not each bundle supports it, here or in README.

File: GPRI.ecl [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    * @params x input data for prediction.
+    * @return rst prediction result
+    * @see ML_Core.Types.NumericField
+  */
+  EXPORT DATASET(NumericField) predict(INTEGER session,
+                                DATASET(Layout_model2) mod, DATASET(NumericField) x) := FUNCTION
+      disMod := DISTRIBUTE(mod, ALL);
+      disX := SORT(DISTRIBUTE(x, id), id, number, LOCAL);
+      rst := rffGPR.predict(disMod, disX, session);
+      RETURN SORT(rst, id);
```
Comment (RogerDev) [inline GPRI.ecl]: Should sort by number as well, I think.

File: GPRI.ecl (hunk 1 above)
Comment (RogerDev) [inline GPRI.ecl]: Leave out return variable name on @return. Where is GPR model defined. Should be noted here.

**Review by RogerDev — CHANGES_REQUESTED** [review]
Almost there. A few comments, mostly regarding doc.

**Comment by lilyclemson** [comment]
@RogerDev  changes are made based on the comments. Please review.

**Review by RogerDev — APPROVED** [review]
@ghalliday @lilyclemson   This is good to merge

File: GPRI.ecl [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+NumericField := MTypes.NumericField;
+
+/* Random Fourier Features accelerated Gaussian Process Regressor
+ *
+ * Random Fourier Features(RFF) map the input data to a randomized low-dimensional feature space.
+ * Then one can apply fast existing linear methods to such new space and thus accelerate the
+ * training of large scale kernel machines[1]. This bundle is the accelerated version of Gaussian
+ * Process Regression(GPR) using such random fourier features.
+ *
+ * Three functoins are available for the users: getSession, fit and predict.
```
Comment (ghalliday) [inline GPRI.ecl]: typo: functions

File: GPRI.ecl [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+ *   * GetSession function generates a 'session ID' for the training and predict process.
+ *   * Fit function fits the input data and train a GPR model.
+ *   * Predict funcion uses the trained GPR model to make predictions for the new observations.
```
Comment (ghalliday) [inline GPRI.ecl]: typo: function

File: README.md [hunk 9]
```diff
@@ -1 +1,49 @@
...
+This bundle provides a Random Fourier Features accelerated version of Gaussian Process Regressor.
+ It allows Data Scientist, researchers or software programmers to apply Gaussian Process Regressor
+in the parallelized environment of HPCC Systems.
+
+Random Fourier Features(RFF) map the input data to a randomized low-dimensional feature space.
+Then one can apply fast existing linear methods to such new space and thus accelerate the
+training of large scale kernel machines[1]. This bundle is the accelerated version of Gaussian
+Process Regression(GPR) using such random fourier features.
+
+The module GPRI is the main ECL interface. Three functoins are available to the users: getSession,
```
Comment (ghalliday) [inline README.md]: same typos of "functions"

File: README.md [hunk 10]
```diff
@@ -1 +1,49 @@
...
+to define session ID, fit a GPR model and make predictions. The testing data is generated by
+M_dataGen.ecl file which generates random test data by user defined size.
+
+## OTHER DOCUMENTATIONS
+[HPCC Systems Machine Learning Library](https://hpccsystems.com/download/free-modules/machine-learning-library)
+
+[Using HPCC Systems Machine Learning Library](https://hpccsystems.com/blog/HPCC-Sytems-Machine-Learning)
+
+## ACKNOWLEDGEMENT
+This bunel is built upon the orignal python implementaton of GPR module from below source:
```
Comment (ghalliday) [inline README.md]: typo: bundle

File: Test/test.ecl [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+IMPORT $ as rff;
+
+// Generate test data
+dataGen:= rff.M_dataGen(1000, 800);
+dx_train := DISTRIBUTE(dataGen.x_train, 0): GLOBAL;
+dy_train := DISTRIBUTE(dataGen.y_train, 0): GLOBAL;
+dx_test := DISTRIBUTE(dataGen.x_test, 0): GLOBAL;
+dy_test := DISTRIBUTE(dataGen.y_test, 0): GLOBAL;
+
+// Python versoin GPR
```
Comment (ghalliday) [inline Test/test.ecl]: typo: version

**Review by ghalliday — APPROVED** [review]
@lilyclemson a few minor typos, I would merge if you don't mind, or you can fix first

**Comment by lilyclemson** [comment]
@ghalliday I fixed the typos. Please review.

**Pull request merged**